File: Client/C#/Client/ConnectionManager.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Windows.Forms;
using Google.Protobuf;
using System.Collections.Generic;
using System.Drawing;

namespace Client
{
    public class ConnectionManager
    {
        private readonly IPAddress IP;
        private readonly int port;
        private readonly Client client;
        private readonly Connection connection;
        private LoginForm loginForm;
        private Thread receiveThread;
        private static bool connected;
        private bool serverConnection;

        public ConnectionManager(IPAddress _IP, int _port, Client _client)
        {
            IP = _IP;
            port = _port;
            client = _client;
            connected = false;
            connection = new Connection(IP, port);
        }
        public void Connect()
        {
            receiveThread = new Thread(ReceiveThreadMethod);
            try
            {
                int connResult = connection.Connect();
                if (connResult != 0)
                {
                    Disconnect(false);
                    if (connResult == -1) // max connections reached, couldn't connect
                        throw new SocketException();
                    else if (connResult == -2)
                        throw new Exception("Canceled");
                }
                connected = serverConnection = true;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public void Disconnect(Boolean wasConnected)
        {
            try
            {
                connection.End(wasConnected);
            }
            catch (Exception ex)
            {
                MessageBox.Show("Disconnect error: " + ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
            connected = serverConnection = false;
        }
        public void GroupAction(ClientMessage.Types.groupActionTypes type, String groupName)
        {
            ClientMessage msg = new ClientMessage
            {
                GroupActionType = type,
                MessageType = ClientMessage.Types.messageTypes.Group,
                GroupName = groupName
            };
            SerializeAndSend(msg);
        }
        public void SendMessage(String buffer, String groupName)
        {
            ClientMessage msg = new ClientMessage
            {
                GroupActionType = ClientMessage.Types.groupActionTypes.Message,
                MessageType = ClientMessage.Types.messageTypes.Group,
                MessageContent = buffer,
                GroupName = groupName
            };
            SerializeAndSend(msg);
        }
        public void SendLoginRequest(String _login, String _password)
        {
            ClientMessage user = new ClientMessage
            {
                Login = _login,
                Password = _password,
                AuthorizationType = ClientMessage.Types.authorizationTypes.LogIn,
                MessageType = ClientMessage.Types.messageTypes.Authorization
            };
            SerializeAndSend(user);
        }
        private void SendReply()
        {
            ClientMessage msg = new ClientMessage
            {
                MessageType = ClientMessage.Types.messageTypes.Reply
            };
            SerializeAndSend(msg);
        }
        public void SerializeAndSend(ClientMessage msg)
        {
            byte[] serialized = msg.ToByteArray();
            int msgSize = serialized.Length;
            byte[] result = BitConverter.GetBytes(msgSize);

            byte[] message = new byte[result.Length + msgSize];
            Buffer.BlockCopy(result, 0, message, 0, result.Length);
            Buffer.BlockCopy(serialized, 0, message, result.Length, msgSize);
            int bytesToSend = message.Length;
            while (bytesToSend != 0)
            {
                int sent = connection.Socket.Send(message, bytesToSend, SocketFlags.None);
                bytesToSend -= sent;
            }
        }
        private void ReceiveThreadMethod()
        {
             while (connected)
            {
                int received = 0, toRead = 4, bytesRead = 0;
                byte[] answer = new byte[4];
                List<byte> answerFull = new List<byte>();
                while (toRead != 0)
                {
                    if ((received = connection.Receive(answer, 0, toRead)) == -1)
                        break;
                    toRead -= received;
                }
                if (received == -1) break;
                int answerSize = BitConverter.ToInt32(answer, 0);
                byte[] answerMsg = new byte[answerSize];

                List<byte> list = new List<byte>();
                while (answerSize != 0)
                {
                    if((received = connection.Receive(answerMsg, 0, answerSize)) == -1)
                        break;
                    foreach (byte i in answerMsg)
                    {
                        list.Add(i);
                    }
                    Array.Clear(answerMsg, 0, answerMsg.Length);
                    answerSize -= received;
                    bytesRead += received;
                }
                if (received == -1) break;
                ClientMessage response = ClientMessage.Parser.ParseFrom(list.ToArray());
                if (bytesRead > 0)
                {
                    HandleResponse(response);
                }
            }
            SendNoConnection();
        }
        private void SendNoConnection()
        {
            if (serverConnection)
            {
                client.Invoke((MethodInvoker)delegate
                {
                    MessageBox.Show("Connection to server lost. Closing client now!", "Connection error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    client.Close();
                });
            }
        }
        private void LoginUser(Boolean correctData, ClientMessage response)
        {
            loginForm.Invoke((MethodInvoker)delegate
                {
                    if (correctData) {
                        InitializeGroups(response);
                        loginForm.AcceptLogin();
                    }
                    else
                    {
                        loginForm.RejectLogin(response.ReplyContent);
                    }
                }
            );
        }
        private void InitializeGroups(ClientMessage respone)
        {
            foreach (String group in respone.Groups)
            {
                client.Groups.Items.Add(group);
            }
        }

        private void HandleResponse(ClientMessage response)
        {
            if (response.Reply == ClientMessage.Types.replyStatus.Positive)
            {
                if (response.AuthorizationType == ClientMessage.Types.authorizationTypes.LogIn)
                    LoginUser(true, response);
                else if (response.GroupActionType != ClientMessage.Types.groupActionTypes.Nogrouptype)
                    GroupReply(response.GroupName, response.GroupActionType);
            }
            else if (response.Reply == ClientMessage.Types.replyStatus.Negative)
            {
                if(response.GroupName.Length != 0 && response.ReplyContent.Length != 0 
                    && response.MessageType == ClientMessage.Types.messageTypes.Reply)
                        GroupReply(response.GroupName, ClientMessage.Types.groupActionTypes.Message);
                if (response.AuthorizationType == ClientMessage.Types.authorizationTypes.LogIn)
                    LoginUser(false, response);
            }
            if (response.MessageType == ClientMessage.Types.messageTypes.Group)
            {
                if (response.GroupActionType == ClientMessage.Types.groupActionTypes.Request)
                    CreateJoinGroupAlert(response.UserName, response.GroupName);
                else if (response.GroupActionType == ClientMessage.Types.groupActionTypes.Accept)
                    GroupReply(response.GroupName, response.GroupActionType);
                else if (response.GroupActionType == ClientMessage.Types.groupActionTypes.Decline)
                    GroupReply(response.GroupName, response.GroupActionType);
            }
            if (response.MessageType != ClientMessage.Types.messageTypes.Reply)
            {
                SendReply();
            }
            if (response.MessageContent.Length != 0)
            {
                client.ChatText.SelectionColor = Color.Black;
                client.ChatText.AppendText(response.UserName + ": " + response.MessageContent + Environment.NewLine);
            }
            if (response.ReplyContent.Length != 0)
            {
                client.ChatText.SelectionColor = Color.Red;
                client.ChatText.AppendText("Error: " + response.ReplyContent + Environment.NewLine);
            }
        }
        private void GroupReply(String groupName, ClientMessage.Types.groupActionTypes reply)
        {
            client.Invoke((MethodInvoker)delegate
            {
                switch (reply)
                {
                    case ClientMessage.Types.groupActionTypes.Create:
                        {
                            client.ChatText.SelectionColor = Color.Green;
                            client.ChatText.AppendText("Group: '" + groupName + "' created!" + Environment.NewLine);
                            client.Groups.Items.Add(groupName);
                            break;
                        }
                    case ClientMessage.Types.groupActionTypes.Delete:
                        {
                            client.ChatText.SelectionColor = Color.Green;
                            client.ChatText.AppendText("Group: '" + groupName + "' deleted!" + Environment.NewLine);
                            client.Groups.Items.Remove(groupName);
                            break;
                        }
                    case ClientMessage.Types.groupActionTypes.Leave:
                        {
                            client.ChatText.SelectionColor = Color.Blue;
                            client.ChatText.AppendText("Group: '" + groupName + "' left!" + Environment.NewLine);
                            client.Groups.Items.Remove(groupName);
                            break;
                        }
                    case ClientMessage.Types.groupActionTypes.Accept:
                        {
                            client.ChatText.SelectionColor = Color.Green;
                            client.ChatText.AppendText("Request accepted. Joined group: " + groupName + Environment.NewLine);
                            client.Groups.Items.Add(groupName);
                            break;
                        }
                    case ClientMessage.Types.groupActionTypes.Decline:
                        {
                            client.ChatText.SelectionColor = Color.Red;
                            client.ChatText.AppendText("Request to join group: " + groupName + " rejected!" + Environment.NewLine);
                            break;
                        }
                    case ClientMessage.Types.groupActionTypes.Request:
                        {
                            client.ChatText.SelectionColor = Color.Blue;
                            client.ChatText.AppendText("Request to join group: '" + groupName + "' send!" + Environment.NewLine);
                            break;
                        }
                    case ClientMessage.Types.groupActionTypes.Message:
                        {
                            client.Groups.Items.Remove(groupName);
                            break;
                        }
                }
            }
            );
        }
        private void CreateJoinGroupAlert(String username, String groupName)
        {
            client.Invoke((MethodInvoker)delegate
            {
                ClientMessage.Types.groupActionTypes type = ClientMessage.Types.groupActionTypes.Decline;
                DialogResult dialog = MessageBox.Show("User: '" + username + "' wants to join group: '" + groupName + "'\nAccept?", "Join request", MessageBoxButtons.YesNo, MessageBoxIcon.Information);
                if (dialog == DialogResult.Yes)
                {
                    type = ClientMessage.Types.groupActionTypes.Accept;
                }
                ClientMessage msg = new ClientMessage
                {
                    MessageType = ClientMessage.Types.messageTypes.Group,
                    GroupActionType = type,
                    UserName = username,
                    GroupName = groupName
                };
                SerializeAndSend(msg);
            });
        }
        public void CheckUser(String _login, String _password, LoginForm loginForm)
        {
            SendLoginRequest(_login, _password);
            this.loginForm = loginForm;       
        }
        public Thread ReceiveThread
        {
            get { return receiveThread; }
        }
    }
    
}

File: Client/Java/src/client/Model/ConnectionManager.java
package client.Model;

import client.Controller.ClientViewController;
import client.Controller.LoginViewController;
import client.Main;
import client.Message;
import com.google.protobuf.InvalidProtocolBufferException;
import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import javafx.scene.control.Alert;
import javafx.scene.control.ButtonType;
import java.io.IOException;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Optional;

public class ConnectionManager {

    private Connection connection;
    private String IP;
    private Integer port;
    private Boolean connected;
    private volatile Boolean serverConnection;
    private ClientViewController client;
    private ObservableList<String> groups;
    private Thread receiveThread;
    private LoginViewController loginController;

    public ConnectionManager(String _ip, Integer _port, ClientViewController clientViewController) {
        IP = _ip;
        port = _port;
        client = clientViewController;
        connected = false;
        connection = new Connection(IP, port);
    }

    public void connect() throws Exception {
        Task work = new Task<Void>(){
            @Override
            protected Void call() {
                ReceiveThreadMethod();
                return null;
            }
        };
        receiveThread = new Thread(work);
        int connResult = connection.connect();
        if (connResult != 0) {

            if (connResult == -1) // max connections reached, couldn't connect
                throw new SocketException();
            if (connResult == -2) // connection canceled by user
                throw new Exception("Canceled");
        }
        connected = serverConnection = true;
    }

    public void disconnect(){
        Main.newAlert(Alert.AlertType.INFORMATION, "Disconnection", "You were disconnected").showAndWait();
        serverConnection = false;
        try {
            connection.end();
        } catch (IOException e) {
            Main.newAlert(Alert.AlertType.ERROR, "Disconnection error!", e.getMessage())
                    .showAndWait();
        }
        connected = false;
    }

    public void groupAction(Message.ClientMessage.groupActionTypes type, String groupName){
        Message.ClientMessage group = Message.ClientMessage.newBuilder()
                .setMessageType(Message.ClientMessage.messageTypes.GROUP)
                .setGroupActionType(type)
                .setGroupName(groupName)
                .build();

        serializeAndSend(group);
    }

    public void sendMessage(String buffer, String groupName){

        Message.ClientMessage msg = Message.ClientMessage.newBuilder()
                .setGroupActionType(Message.ClientMessage.groupActionTypes.MESSAGE)
                .setMessageType(Message.ClientMessage.messageTypes.GROUP)
                .setMessageContent(buffer)
                .setGroupName(groupName)
                .build();

        serializeAndSend(msg);
    }

    private void sendLoginRequest(String login, String password){
        Message.ClientMessage user = Message.ClientMessage.newBuilder()
                .setLogin(login).setPassword(password)
                .setAuthorizationType(Message.ClientMessage.authorizationTypes.LOG_IN)
                .setMessageType(Message.ClientMessage.messageTypes.AUTHORIZATION)
                .build();

        serializeAndSend(user);
    }

    private void sendReply(){
        Message.ClientMessage msg = Message.ClientMessage.newBuilder()
                .setMessageType(Message.ClientMessage.messageTypes.REPLY)
                .build();

        serializeAndSend(msg);
    }

    private void serializeAndSend(Message.ClientMessage msg){
        byte[] serialized = msg.toByteArray();
        int msgSize = serialized.length;
        byte[] result = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(msgSize).array();

        byte[] message = new byte[result.length + msgSize];
        System.arraycopy(result, 0, message, 0, result.length);
        System.arraycopy(serialized, 0, message, result.length, msgSize);

        connection.send(message);
    }

    private int getMsgLength(char[] answer){
        byte[] receivedBytes = new String(answer).getBytes();
        ByteBuffer wrapped = ByteBuffer.wrap(receivedBytes);
        wrapped.order(ByteOrder.LITTLE_ENDIAN);
        return wrapped.getInt();
    }

    private void ReceiveThreadMethod() {

        while(connected){
            int received, toRead = 4, bytesRead = 0;
            char[] answer = new char[4];
            while(toRead != 0) {
                if ((received = connection.receive(answer, 0, 4)) == -1) {
                    break;
                }
                toRead -= received;
            }
            int answerSize = getMsgLength(answer);
            char[] answerMsg = new char[answerSize];
            toRead = answerSize;
            while(toRead != 0) {
                if((received = connection.receive(answerMsg, 0, answerSize)) == -1){
                    break;
                }
                toRead -= received;
                bytesRead += received;
            }

            Message.ClientMessage response = null;
            try {
                response = Message.ClientMessage.parseFrom(new String(answerMsg).getBytes());
            } catch (InvalidProtocolBufferException e) {
                e.printStackTrace();
            }

            if(bytesRead > 0){
                assert response != null;
                handleResponse(response);
            }
        }
        sendNoConnection();
    }

    private void sendNoConnection(){
        Thread t = new Thread(() -> Platform.runLater(() -> {
            if(serverConnection) {
                Main.newAlert(Alert.AlertType.INFORMATION, "Disconnection", "Server connection lost! You were disconnected").showAndWait();
                try {
                    connection.end();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            client.openConnectForm();
        }));
        t.start();
    }

    private void loginUser(Boolean correctData, Message.ClientMessage response){
        Thread t = new Thread(() -> Platform.runLater(() -> {
            if(correctData){
                ObservableList<String> groups = FXCollections.observableArrayList(response.getGroupsList());
                client.getGroupChoice().setItems(groups);
                loginController.acceptLogin();
            }
            else{
                loginController.rejectLogin(response.getReplyContent());
            }
        }));
        t.start();
    }

    private void handleResponse(Message.ClientMessage response){
        if(response.getReply() == Message.ClientMessage.replyStatus.POSITIVE) {
            if(response.getAuthorizationType() == Message.ClientMessage.authorizationTypes.LOG_IN)
                loginUser(true, response);
            else if(response.getGroupActionType() != Message.ClientMessage.groupActionTypes.NOGROUPTYPE)
                groupReply(response.getGroupName(), response.getGroupActionType());
        }
        else if(response.getReply() == Message.ClientMessage.replyStatus.NEGATIVE){
            if(!response.getGroupName().isEmpty() && !response.getReplyContent().isEmpty() && response.getMessageType() == Message.ClientMessage.messageTypes.REPLY) {
                groupReply(response.getGroupName(), Message.ClientMessage.groupActionTypes.MESSAGE);
            }
            if(response.getAuthorizationType() == Message.ClientMessage.authorizationTypes.LOG_IN)
                loginUser(false, response);
        }
        if(response.getMessageType() == Message.ClientMessage.messageTypes.GROUP){
            if(response.getGroupActionType() == Message.ClientMessage.groupActionTypes.REQUEST)
                createJoinGroupAlert(response.getUserName(), response.getGroupName());
            else if(response.getGroupActionType() == Message.ClientMessage.groupActionTypes.ACCEPT)
                groupReply(response.getGroupName(), response.getGroupActionType());
            else if(response.getGroupActionType() == Message.ClientMessage.groupActionTypes.DECLINE)
                groupReply(response.getGroupName(), response.getGroupActionType());
        }
        if(response.getMessageType()!= Message.ClientMessage.messageTypes.REPLY)
            sendReply();

        if(!response.getMessageContent().isEmpty())
            client.getTextArea().appendText(response.getUserName() +  ": " + response.getMessageContent() + '\n');

        if(!response.getReplyContent().isEmpty() && response.getAuthorizationType() != Message.ClientMessage.authorizationTypes.LOG_IN)
            client.getTextArea().appendText("Error: " + response.getReplyContent() + '\n');
    }

    private void groupReply(String groupName, Message.ClientMessage.groupActionTypes type){
        Platform.runLater(()-> {
            ObservableList newList = client.getGroupChoice().getItems();
            switch(type) {
                case CREATE: {
                    client.getTextArea().appendText("Group: '" + groupName + "' created!" + '\n');
                    newList.add(groupName);
                    break;
                }
                case DELETE: {
                    client.getTextArea().appendText("Group: '" + groupName + "' deleted!" + '\n');
                    newList.remove(groupName);
                    break;
                }
                case LEAVE: {
                    client.getTextArea().appendText("Group: '" + groupName + "' left!" + '\n');
                    newList.remove(groupName);
                    break;
                }
                case ACCEPT: {
                    client.getTextArea().appendText("Request accepted. Joined group: '" + groupName + "'" + '\n');
                    newList.add(groupName);
                    break;
                }
                case DECLINE: {
                    client.getTextArea().appendText("Request to join group: '" + groupName + "' rejected!\n");
                    break;
                }
                case REQUEST: {
                    client.getTextArea().appendText("Request to join group: '" + groupName + "' send!\n");
                    break;
                }
                case MESSAGE: {
                    newList.remove(groupName);
                    break;
                }
            }
            client.getGroupChoice().setItems(newList);
            client.getGroupChoice().getSelectionModel().clearSelection();
        });
    }

    private void createJoinGroupAlert(String userName, String groupName){

        Thread t = new Thread(() -> Platform.runLater(() -> {
            Alert alert = Main.newAlert(Alert.AlertType.CONFIRMATION, "Request", "User: " + userName +
                    " wants to join group: " + groupName);
            ButtonType buttonAccept = new ButtonType("Accept"),
                        buttonDecline = new ButtonType("Decline");
            alert.getButtonTypes().setAll(buttonAccept, buttonDecline);
            Optional<ButtonType> option = alert.showAndWait();

            Message.ClientMessage.groupActionTypes type = Message.ClientMessage.groupActionTypes.DECLINE;
            if (option.isPresent() && option.get() == buttonAccept) {
                type = Message.ClientMessage.groupActionTypes.ACCEPT;
            }
            Message.ClientMessage msg = Message.ClientMessage.newBuilder()
                    .setMessageType(Message.ClientMessage.messageTypes.GROUP)
                    .setGroupActionType(type)
                    .setUserName(userName)
                    .setGroupName(groupName)
                    .build();

            serializeAndSend(msg);
        }));
        t.start();
    }
    public void checkUser(String login, String password, LoginViewController _loginController) throws IOException {
        loginController = _loginController;
        sendLoginRequest(login, password);
    }

    public Thread getReceiveThread() { return receiveThread; }
    public Boolean isConnected(){
        return connected;
    }
    public ObservableList<String> getGroups() { return groups; }
}
